src/browser/hooks/useAnimationEnd.js

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+
+function useAnimationEnd(
+  ref,
+  callback,
+  animationName,
+  allowEventBubbling = true,
+) {
+  React.useEffect(() => {
+    if (!ref.current) {
+      return;
```

Comment:
if ref changed and current is no longer available, should we remove the listener? or can it be reasigned?

Reply:
The final returned arrow function in the `useEffect` closure function is used to unregister the event listener appropriately. Hooks are very interesting in how they work! Because of the array of props provided to the `useEffect` call, if the ref changes (or any of the other provided props), the returned arrow function will automatically be run (thereby unregistering the event handler) before the `useEffect` handler function is rerun to  again register the event listener on the changed ref. If `ref.current` no longer exists, nothing will be registered because of the initial conditionals in the `useEffect` closure function. If none of the provide props change, the `useEffect` closure function does not get re-run.

Reply:
Yes, your PR made me refresh what I read about hooks, and you are correct that it will run the functions on `ref` changing, but the function returned first checks if it's null and then returns (I should probably have added the comment to that second check instead of this one 😓 see line 30, sorry for the confusion)

Reply:
Right. Honestly I thought the same at first and didn't have the conditional, but ran into an error somewhere along the line where ref.current was null at that point so there was nothing to remove. That said, it must have been something else, because an arrow function will never be returned if ref.current is null to begin with so you are of course correct – good catch! :)